"""
Author: Ali Saadat
Affiliation: EPFL
Aim: A Snakemake workflow to process paired-end sequencing data (WGS) using bwa/GATK4. 
Date created: 16-May-2022
Adapted from https://github.com/ESR-NZ/human_genomics_pipeline
"""

##### Set up wildcards #####

# Define samples from fastq dir and chromosmes
SAMPLES, = glob_wildcards("../data/fastq/{sample}_R1.fastq.gz")
CHROMS = [str(c) for c in range(1, 23)] + ["X", "Y", "M"]

##### Setup helper functions #####
import csv
import glob

def get_recal_resources_command(resource):
    command = ""
    for resource in config['RECALIBRATION']['RESOURCES']:
        command += "--known-sites " + resource + " "
    return command

def get_pedigree_command(resource):
    command = ""
    if config['PED'] != "":
        command = "-ped " + config['PED']
    return command

def get_vqsr_DP_option(resource):
    if config['DATA'] == "WES":
        command = ""
    if config['DATA'] == "WGS":
        command = "-an DP"
    return command

def get_vqsr_InbreedingCoefficient_option(resource):
    if config['INBREED_COEFF_FILTER'] == "EXCLUDE":
        command = ""   
    if config['INBREED_COEFF_FILTER'] == "INCLUDE":
        command = "-an InbreedingCoeff"
    return command

def get_wes_intervals_command(resource):    
    command = ""
    command = "-L " + config['WES']['PADDING'] + " "
    if config['WES']['INTERVALS'] == "":
        command = ""
    return command

def get_wes_padding_command(resource):
    command = ""
    command = "--ip " + config['WES']['PADDING'] + " " 
    if config['WES']['PADDING'] == "":
        command = ""
    return command

def get_bwa_memory(resource):
    if config['DATA'] == "WES":
        return 15000
    if config['DATA'] == "WGS":
        return 50000
    else:
        return 15000

def get_mkdup_memory(resource):
    if config['DATA'] == "WES":
        return 40000
    if config['DATA'] == "WGS":
        return 80000
    else:
        return 40000

def get_mkdup_xmx(resource):
    if config['DATA'] == "WES":
        expand('"-Xmx{maxmemory}"', maxmemory = config['MAXMEMORY']['MARK_DUP_WES'])
    if config['DATA'] == "WGS":
        expand('"-Xmx{maxmemory}"', maxmemory = config['MAXMEMORY']['MARK_DUP_WGS'])
    else:
        expand('"-Xmx{maxmemory}"', maxmemory = config['MAXMEMORY']['MARK_DUP_WES'])

def get_HC_memory(resource):
    if config['DATA'] == "WES":
        return 10000
    if config['DATA'] == "WGS":
        return 20000
    else:
        return 10000

def get_HC_xmx(resource):
    if config['DATA'] == "WES":
        expand('"-Xmx{maxmemory}"', maxmemory = config['MAXMEMORY']['HC_WES'])
    if config['DATA'] == "WGS":
        expand('"-Xmx{maxmemory}"', maxmemory = config['MAXMEMORY']['HC_WGS'])
    else:
        expand('"-Xmx{maxmemory}"', maxmemory = config['MAXMEMORY']['HC_WES'])
        
#### Set up report #####

report: "report/workflow.rst"

##### Target rules depnding on the existense of pedigree file #####
localrules: all 

if config['PED'] == "":
    rule all:
        input:
            expand("../results/qc/fastqc/{sample}_R{read}_fastqc.html", sample = SAMPLES, read = [1, 2]),
            "../results/vcf/refined_GQ.vcf.gz"
if config['PED'] != "":
    rule all:
        input:
            expand("../results/qc/fastqc/{sample}_R{read}_fastqc.html", sample = SAMPLES, read = [1, 2]),
            "../results/vcf/refined_GQ_denovo.vcf.gz"

##### Load rules #####

include: "rules/00_fastqc.smk"
include: "rules/01_bwa_mem.smk"
include: "rules/02_gatk_MarkDuplicates.smk" 
include: "rules/03_gatk_BaseRecalibrator.smk"
include: "rules/04_gatk_ApplyBQSR.smk"
include: "rules/05_gatk_HaplotypeCaller_cohort.smk"
include: "rules/06_gatk_GenomicsDBImport.smk" 
include: "rules/07_gatk_GenotypeGVCFs.smk"
include: "rules/08_gatk_GatherVCF.smk" 

if config['FILTER'] == "VQSR": 
    include: "rules/09_gatk_VQSR.smk"

if config['FILTER'] == "HARD": 
    include: "rules/09_gatk_HARD.smk"

include: "rules/10_gatk_GenotypeRefinement.smk"

if config['PED'] != "":
    include: "rules/11_gatk_denovo.smk"